ant-networking/src/lib.rs

```diff
@@ -406,54 +406,51 @@ impl Network {
             .send_and_get_responses(&close_nodes, &request, true)
             .await;
 
-        // loop over responses, generating an average fee and storing all responses along side
-        let mut all_costs = vec![];
+        // loop over responses
         let mut all_quotes = vec![];
-        for response in responses.into_values().flatten() {
+        let mut quotes_to_pay = vec![];
+        for (peer, response) in responses {
             info!(
-                "StoreCostReq for {record_address:?} received response: {:?}",
-                response
-            );
+                "StoreCostReq for {record_address:?} received response: {response:?}");
             match response {
-                Response::Query(QueryResponse::GetStoreCost {
+                Ok(Response::Query(QueryResponse::GetStoreQuote {
                     quote: Ok(quote),
-                    payment_address,
                     peer_address,
                     storage_proofs,
-                }) => {
+                })) => {
                     if !storage_proofs.is_empty() {
-                        debug!("Storage proofing during GetStoreCost to be implemented.");
+                        debug!("Storage proofing during GetStoreQuote to be implemented.");
                     }
                     // Check the quote itself is valid.
-                    if quote.cost
-                        != AttoTokens::from_u64(calculate_cost_for_records(
-                            quote.quoting_metrics.close_records_stored,
-                        ))
-                    {
+                    if !quote.check_is_signed_by_claimed_peer(peer) {
                         warn!("Received invalid quote from {peer_address:?}, {quote:?}");
                         continue;
                     }
 
-                    all_costs.push((peer_address.clone(), payment_address, quote.clone()));
-                    all_quotes.push((peer_address, quote));
+                    all_quotes.push((peer_address.clone(), quote.clone()));
+                    quotes_to_pay.push((peer, quote));
                 }
-                Response::Query(QueryResponse::GetStoreCost {
+                Ok(Response::Query(QueryResponse::GetStoreQuote {
                     quote: Err(ProtocolError::RecordExists(_)),
-                    payment_address,
                     peer_address,
                     storage_proofs,
-                }) => {
+                })) => {
                     if !storage_proofs.is_empty() {
-                        debug!("Storage proofing during GetStoreCost to be implemented.");
+                        debug!("Storage proofing during GetStoreQuote to be implemented.");
                     }
-                    all_costs.push((peer_address, payment_address, PaymentQuote::zero()));
+                    info!("Address {record_address:?} was already paid for according to {peer_address:?}, ending quote request");
+                    return Ok(vec![]);
```

Comment:
this actually creates a hole that if any malicious node replies with RecordExists , it will blocks the upload.
I'd suggest leave it and continue with collect of other responses.
Then decided whether to pay or not by: if find majority of RecordExists responses, or verified can fetch from network